src/client/Client.cc

```diff
@@ -3755,8 +3756,13 @@ void Client::send_cap(Inode *in, MetaSession *session, Cap *cap,
 				   flush,
 				   cap->mseq,
                                    cap_epoch_barrier);
-  m->caller_uid = in->cap_dirtier_uid;
-  m->caller_gid = in->cap_dirtier_gid;
+  /*
+   * Since the setattr will check the cephx mds auth access before
+   * buffering the changes, so it makes no sense any more to let
+   * the cap update to check the access in MDS again.
```

Comment:
What about a newer client talking to an old MDS (i.e. one without `CEPHFS_FEATURE_MDS_AUTH_CAPS_CHECK`)?

I think we're basically matching the existing kernel behavior, right? It's important to explain here in a comment what this is about. Also link to the tracker ticket you're fixing.

Reply:
For the newer clients taking an old MDS case it will be fixed in https://github.com/ceph/ceph/pull/47506 by @ajarr. @ajarr's commits will disallow the newer clients to mount if the `root_squash` enabled and the MDS doesn't support `CEPHFS_FEATURE_MDS_AUTH_CAPS_CHECK`. 

I think it'd better pull @ajarr's following commits to this PR to make it's more easy to review and test.

Reply:
I think I missed understanding this.

@ajarr's PR will just fix the old client without supporting `CEPHFS_FEATURE_MDS_AUTH_CAPS_CHECK` case.

For this case I will add one commit to force the session to be readonly in client side instead.

Reply:
PR #47506 has commit from this PR, so its a bit confusing. Can you pull the relevant commits from that PR into this @lxbsz ?

Reply:
Yeah, already done that. I have pulled @ajarr's 4 commits into this one.